### ISSUE TITLE
Increase functionalities on the `Strings` module

### ISSUE THREAD

**Issue by teofr** [mulesoft-labs/data-weave-rfc#25]
The dw::core::Strings module has many useful functions for dealing semantically with strings (`pluralize`, `camelize`, ...)
but lacks functions for dealing with a String as if it were a sequence of characters.

The proposal is to add the following functions to the Strings module, to ease working with them.

#### `map(String, (item: String, index: Number) -> String): String`

Notice that, since we don't differentiate between String and characters, `map` would allow to delete or add characters,
unless we fail explicitely (should we?)

```
"hola" map $ + "!"  =>  "h!o!l!a!"
```

#### `splitBy(String, (item: String, index: Number) -> Boolean): Array<String>`

```
"hello world!" splitBy isVowel($) => ["h", "ll", " w", "rld!"]
```

#### `words(String): Array<String>`

```
words("hello world!\nline")  =>  ["hello", "world!", "line"]
```

#### `lines(String): Array<String>`

```
words("hello world!\nline")  =>  ["hello world!", "line"]
```

#### `slice(String, Number, Number): String`

```
slice("hello world!", 2, 8)  =>  "llo wo"
```

#### `divideBy(String, Number): Array<String>`

```
divideBy("hello world!", 5)  =>  ["hello", " worl", "d!"]
```

#### `take(String, Number): String`

```
take("hello world!", 4)  =>  "hell"
```

#### `takeWhile(String, (item: T) -> Boolean): String`

```
takeWhile("hello world!", $ != " ")  => "hello"
```

#### `drop(String, Number): String`

```
drop("hello world!", 4)  =>  "o world!"
```

#### `dropWhile(String, (item: T) -> Boolean): String`

```
dropWhile("hello world!", $ != " ")  => " world!"
```

#### `--(String, Array<String>): String`

```
"hello world" -- ["ll", "o"]  =>  "heo wrld!"
```

This may become problematic, since it's not clear which one has precedence, for instance
should `"hello world" -- ["lo", "ll"]` result in `"hel world!"` (deleting `"lo"`) or in
`"heo world!"` (deleting `"ll"`)

As an alternative, having `--(String, String): String` could be good enough.

```
fun --(x: String, y: String) : String = x replace y with ""
```

#### `groupBy(String, (item: String, index: Number) -> R): { (R): String }`

```
"hello world!" groupBy isVowel($)  =>  { true: "eoo", false: "hll wrld!"}
```

#### `countBy(String, (String) -> Boolean): Number`

```
"hello world!" countBy isVowel($)  =>  3
```

#### `filter(String, (item: String, index: Number) -> Boolean): String`

```
"hello world!" filter ((item, index) ->  (index mod 2) == 0)  =>  "hlowrd"
```

#### `toArray(String): Array<String>`
Alternative name: `charsOf`

```
toArray("hello")  =>  ["h", "e", "l", "l", "o"]
```

#### `fromArray(Array<String>): String`
Alternative name: `fromChars`

```
fromArray(["h", "e", "l", "l", "o"])  =>  "hello"
```

#### `firstFind(text: String, needle: Regex): Array<Number`

Index of first appearance of needle in text.

```
"bye bye world" firstFind "bye"  =>  0
```

#### `lastFind(text: String, needle: Regex): Array<Number>`

Index of last appearance of needle in text.

```
"bye bye world" lastFind "bye"  =>  4
```

Note: in the regex cases above, each Array result hast the index as the first element,
and the capture groups in the next ones.

#### `count(text: String, needle: Regex): Number`

Count the number of time `needle` appears on `text`

```
"bye bye world" count "bye"  =>  2
```

**Comment by machaval** [comment]
One thing that we should try to avoid in functions name clashing between our Modules. Most of the time people do `import * from x` and if so then if two modules has same function name it will do name clashing and currently it takes the one from the first imported module. For that reason I would prefer for us to look for other function name for the cases were those function are already taken.

**Comment by teofr** [comment]
> One thing that we should try to avoid in functions name clashing between our Modules. Most of the time people do `import * from x` and if so then if two modules has same function name it will do name clashing and currently it takes the one from the first imported module. For that reason I would prefer for us to look for other function name for the cases were those function are already taken.

Noted, maybe appending `String` to the name, like `sliceString` or `divideByString` (this one is confusing) Or do you think looking for different names is better? `divideBy` -> `split`, `slice` -> `substring`, and so on.

The ones that are on Core are alright, right? (`map`, `reduce`, ..)

**Comment by adobni** [comment]
It would a good idea to add a String implementation of every() and some(), like the Array module has for arrays. The string versions would apply a condition to each character of a string.
```
every(String, (T) -> Boolean): Boolean
count(String, (T) -> Boolean): Boolean
```

**Comment by teofr** [comment]
Just a small update, based on the comments I changed the following functions presented above:

`map(String, (item: String, index: Number) -> String): String`
Written as `flatmap,` since it fits better with the functionality. The rest is the same

`take(String, Number): String`
written as `prefix`, since it fits better with the `Strings` module, and
prevents a name clash.

`drop(String, Number): String`
Removed in favor of `suffix`, the functionality is slightly different.

`suffix("hello!", 2) => "o!"`

`takeWhile(String, (item: T) -> Boolean): String`
`dropWhile(String, (item: T) -> Boolean): String`
Both of this are removed, mainly to avoid adding functions like `dropWhileString`, but
also since they can easily be recovered using `splitBy`.

`everyCharacter(String, (String) -> Boolean): Boolean`
`someCharacter(String, (String) -> Boolean): Boolean`
Added this functions based on @adobni's  suggestion.
